src/monattack.cpp

```diff
@@ -3462,38 +3484,59 @@ void mattack::generator(monster *z, int index)
 void mattack::upgrade(monster *z, int index)
 {
     std::vector<int> targets;
+    int day = calendar::turn / DAYS(1);
     for (size_t i = 0; i < g->num_zombies(); i++) {
         monster &zed = g->zombie(i);
-        if( zed.type->id == "mon_zombie" &&
-            rl_dist( z->pos(), zed.pos() ) <= 5 &&
-            z->sees( zed ) &&
-            z->attitude_to( zed ) != Creature::Attitude::A_HOSTILE ) {
-            targets.push_back(i);
+        // Check this first because it is a cheap check
+        if( zed.type->upgrade_min < day &&
+           (zed.type->upgrade_group != "NULL" || zed.type->upgrades_into != "NULL" )) {
+            // Then do the more expensive ones
+            if ( z->attitude_to( zed ) != Creature::Attitude::A_HOSTILE &&
+                 within_target_range(z, &zed, 10) ) {
+                targets.push_back(i);
+            }
         }
     }
     if (targets.empty()) {
+        // Nobody to upgrade, get MAD!
+        z->anger = 100;
         return;
+    } else {
+        // We've got zombies to upgrade now, calm down again
+        z->anger = 5;
     }
+
     z->reset_special(index); // Reset timer
-    z->moves -= 150;   // It takes a while
+    z->moves -= z->type->speed; // Takes one turn
 
     monster *target = &( g->zombie( targets[ rng(0, targets.size() - 1) ] ) );
 
-    const auto monsters = MonsterGroupManager::GetMonstersFromGroup("GROUP_ZOMBIE_UPGRADE");
-    const std::string newtype = monsters[rng(0, monsters.size() - 1)];
-
+    std::string old_name = target->name();
     const auto could_see = g->u.sees( *target );
-    target->poly(GetMType(newtype));
-    const auto can_now_see = g->u.sees( *target );
+    // Currently gives zombies the equivalent of a week of upgrade time
+    // Difficulty scaling happens in update_check()
+    target->set_last_load(target->get_last_load() - 7);
```

Comment:
At the moment yes, but it may take several uses on each individual zombie to actually get them to that stage.